lib/reek/examiner.rb

```diff
@@ -86,35 +87,31 @@ def smelly?
 
     private
 
-    attr_reader :collector, :source, :smell_repository
+    attr_reader :source, :smell_repository
 
-    # Runs the Examiner on the given source to scan for code smells
-    # and returns the corresponding Examiner instance.
+    # Runs the Examiner on the given source to scan for code smells and returns
+    # the resulting smells.
     #
     # In case one of the smell detectors raises an exception we probably hit a Reek bug.
     # So we catch the exception here, let the user know something went wrong
     # and continue with the analysis.
     #
-    # @return an instance of Examiner
-    #
-    # :reek:TooManyStatements: { max_statements: 6 }
     def run
-      @run ||= begin
-        syntax_tree = source.syntax_tree
-        return self unless syntax_tree
-        begin
-          examine syntax_tree
-        rescue StandardError => exception
-          $stderr.puts format(INCOMPREHENSIBLE_SOURCE_TEMPLATE, source.origin, exception.inspect)
-        else
-          smell_repository.report_on(collector)
-        end
-        self
+      return [] unless syntax_tree
+      begin
+        examine_tree
+      rescue StandardError => exception
+        $stderr.puts format(INCOMPREHENSIBLE_SOURCE_TEMPLATE, origin, exception.inspect)
+        []
       end
     end
 
-    def examine(syntax_tree)
-      ContextBuilder.new(syntax_tree).context_tree.each do |element|
+    def syntax_tree
+      @syntax_tree ||= source.syntax_tree
+    end
+
+    def examine_tree
+      ContextBuilder.new(syntax_tree).context_tree.each.flat_map do |element|
```

Comment:
This looks a little odd, but because `context_tree` is not an `Enumerable`, I cannot call `#flat_map` on it directly. I'm toyed with the idea of including `Enumerable` in `CodeContext`, but I'm not sure it's a good idea since it is so much more than a container.